### ISSUE TITLE
Low latency API

### ISSUE THREAD

**Issue by tguillem** [xiph/rav1e#689]
Hello,

I just read the C API and have a remark.

The current API doesn't allow to use the encoder asynchronously.
What most encoders and rav1e are doing is: queuing input frames and dequeing output packets from the same thread or same loop.

Queuing input frames is generally paced by I/O. If your encoder has an output packet ready but the application using your API is stuck in I/O, the output packet will be sent with a high delay.

To solve this issue, the encoder should either provide a callback, that is called when an output packet is encoded. In that case, one application could process the output packet directly from this callback.

An other way to solve this is issue is to make your queuing and dequeing functions thread-safe (rav1e_send_frame() and rav1e_receive_packet()). That way, one application could wait for output packets from their own thread via rav1e_receive_packet() (you should then use a condition variable somewhere). You should also provide a way to cancel the previous call, in order to interrupt the output thread without delays.

**Comment by ycho** [comment]
Yes, I agree that rav1e needs this better way of handling input frames to it and output bitstream from it asynchronously. Thank you for pointing this out!

**Comment by vibhoothiiaanand** [comment]
With help of #2118 Channel API, we can achieve something similar :)
